### ISSUE TITLE
Specify the rule for RAML data type to XML conversion

### ISSUE THREAD

**Issue by svacas** [raml-org/raml-spec#377]
The spec should include a section with details about the conversion between RAML data types and XML and what is expected from a parser implementation.

**Comment by usarid** [comment]
Please see my comment on https://github.com/raml-org/raml-spec/issues/379 and clarify here accordingly, and then the question will become clear: how to validate XML-formatted instances of data against a RAML data type.

**Comment by petrochenko-pavel-a** [comment]
Hi @svacas 

As for me transformation should be performed by the following rules:

Check if data type have defined XML facet, if it is not then you should fill one with the defaults which are:
name: equal to the name of the data type
namespace: null (no namespace)
prefix: null (no prefix)
wrapped: false
attribute: false

No lets differentiate behaviour depending from the kind of instance that’s being serialised:

If it is a scalar then you should create top level element with then name taken from namespace, name , prefix. put string representation of the value in the node text content and the job is done

If it is an array then you should should create top level element with then name taken from namespace, name , prefix. Then you should serialise elements of the array following the rules of serialisation for the component type

If it is an object then you should create top level element with then name taken from namespace, name , prefix. Then you should take every defined property of the object and serialise them by the following algorithm:

	a) if the property is scalar and attribute is true, then you should create an attribute with appropriate way, put string representation of the value inside of it.
	b) if the property is object or scalar with attribute equal to false (default) then you should serialise it just by applying exactly same algorithm to it. 
	c) If the property is array, then if wrapped is false you should take each element of the array and create an element for it. other wise you should apply the algorithm recursively

If the name of the element or the attribute is not compliant to the XML naming convention then you should emulate this name by the pair of elements one for the key, the other for the value following each other sequentially with names "_key" and “_value” ,”_”, (same if name is equal to _key or to _value)

for example you have a following raml type:

X:
  type: object
  properties: 
    1: number

then sample representation of instance in xml might look like:

<X>
   <_>
   <key>1</key>
   <value>some value of property</value>
   </_>)
</X>

P.S. please let me know if you see something inconsistent and let's work together on resolving it.

Regards,
Pavel

**Comment by fededonna** [comment]
the real example should be

``` xml
<X>
   <_key>1</_key>
   <_value>some value of property</_value>
</X>
```

since the tag <_> is used to represent a yaml array and the properties is not an array. Also "key" and "value" should be changed to "_key" and "_value" to specify that the yaml type is not applicable to an xml tag.

**Comment by petrochenko-pavel-a** [comment]
>since the tag <_> is used to represent a yaml array and the properties is not an array
from where you took it?

>Also "key" and "value" should be changed to "_key" and "_value"
Why?

**Comment by fededonna** [comment]
> from where you took it?
from here:
http://yaml.org/xml.html

please check the part of the example that states:

```xml
<product>
      <_>
        <sku>BL394D</sku>
        <quantity>4</quantity>
        <description>Basketball</description>
        <price>450.00</price>
      </_>
      <_>
        <sku>BL4438</sku>
        <quantity>1</quantity>
        <description>Super Hoop</description>
        <price>2392.00</price>
      </_>
   </product>
```

every one of those is a product inside the "product" tag

**Comment by sichvoge** [comment]
@fededonna how would you represent multiple properties if you remove `<_>`? The following case:

```yaml
X:
  type: object
  properties: 
    1: number
    2: string
```

The XML can't probably be:

```xml
<X>
   <_key>1</_key>
   <_value>some value of property</_value>
   <_key>2</_key>
   <_value>some value of property</_value>
</X>
```

But instead (as Pavel said) should be

```xml
<X>
   <_>
      <_key>1</_key>
      <_value>some value of property</_value>
   </_>
   <_>
      <_key>2</_key>
      <_value>some value of property</_value>
   </_>
</X>
```

Right?

**Comment by sichvoge** [comment]
OK, after double checking and applying the rules mentioned in http://yaml.org/xml.html; its indeed 

```xml
<X>
   <_key>1</_key>
   <_value>some value of property</_value>
   <_key>2</_key>
   <_value>some value of property</_value>
</X>
```

@petrochenko-pavel-a; the following rule got a bit modified by you compared to the one after that. Any intention behind that?

Your version:

> If the name of the element or the attribute is not compliant to the XML naming convention then you should emulate this name by the pair of elements one for the key, the other for the value following each other sequentially with names "_key" and “_value” ,””, (same if name is equal to _key or to _value)

Version from http://yaml.org/xml.html:

> Nested keys, typed keys, or keys not matching the XML name production are emulated through a pair of elements, one for the key, the other for the value following each other sequentially with names "_key" and "_value".

I like the second better since it talks about matching YAML keys to XML; yours is already talking about XML only. @fededonna whats your take?

BTW, if we take that rule from the website - we should clearly mention that in the spec.

**Comment by fededonna** [comment]
@sichvoge I think the one in the yaml.org site is the one we should follow. Also, as we discussed before, this should be mentioned in the spec.

**Comment by sichvoge** [comment]
I agree! @petrochenko-pavel-a can you confirm. Where would you add that?

**Comment by sichvoge** [comment]
Should that go into https://github.com/raml-org/raml-spec/blob/raml-10-rc2/versions/raml-10/raml-10.md#xml-serialization-of-type-instances?

**Comment by petrochenko-pavel-a** [comment]
>Should that go into https://github.com/raml-org/raml-spec/blob/raml-10-rc2/versions/raml-10/raml-10.md#xml-serialization-of-type-instances?

yep.
